staging/src/k8s.io/apiserver/pkg/util/x509metrics/server_cert_deprecations.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package x509metrics
+
+import (
+	"crypto/x509"
+	"errors"
+	"net/http"
+	"reflect"
+	"strings"
+
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/component-base/metrics"
+)
+
+var _ utilnet.RoundTripperWrapper = &x509DeprecatedCertificateMetricsRTWrapper{}
+
+type x509DeprecatedCertificateMetricsRTWrapper struct {
+	rt http.RoundTripper
+
+	checkers []deprecatedCertificateAttributeChecker
+}
+
+type deprecatedCertificateAttributeChecker interface {
+	// CheckRoundTripError returns true if the err is an error specific
+	// to this deprecated certificate attribute
+	CheckRoundTripError(err error) bool
+	// CheckPeerCertificates returns true if the deprecated attribute/value pair
+	// was found in a given certificate in the http.Response.TLS.PeerCertificates bundle
+	CheckPeerCertificates(certs []*x509.Certificate) bool
+	// IncreaseCounter increases the counter internal to this interface
+	IncreaseMetricsCounter()
```

Comment:
The metrics reveal that *an* issue exists, but I think we should log something with information about the host being contacted in order for an admin observing these metrics to be able to target a fix. Plumbing the request to this method and logging the host / reason would allow that. Something like this:

```go
if req != nil && req.URL != nil {
  if hostname := req.URL.Hostname(); len(hostname) > 0 {
    klog.Infof("invalid-cert.webhook.kubernetes.io: %s: %s", hostname, c.reason)
  }
}
```

and setting reasons like "relies on legacy Common Name field" or "uses SHA-1 signature" in the specific instances of counterRaiser

Reply:
I'd also consider adding an audit annotation in the same place (which we could do if we have the request context), but that requires audit annotations to be threadsafe (requires something like https://github.com/kubernetes/kubernetes/pull/108716)

Reply:
sounds good